### PR TITLE
Method add_pos for object/player

### PR DESCRIPTION
**Pull request by sfence**
Add the `add_pos` method to Lua API, similar to `add_velocity`.

- Fix issue #14023
- Via Lua API and server to client packet

## To do

This PR is a Ready for Review.

## How to test

The test is included in devtest game unittests command.

### CODE REVIEW

**Comment by Warr1024** [comment]
Works as I expected.

The worldmod I used to test it was:
```lua
local modname = minetest.get_current_modname()

minetest.register_node(modname .. ":north", {
	tiles = { "[combine:16x16^[noalpha^[invert:r" },
	_transport_rel = vector.new(0, 0, 16)
})
minetest.register_node(modname .. ":south", {
	tiles = { "[combine:16x16^[noalpha^[invert:g" },
	_transport_rel = vector.new(0, 0, -16)
})

local function checkplayer(player)
	local pos = vector.offset(player:get_pos(), 0, -0.5, 0)
	local nodename = minetest.get_node(pos).name
	local def = minetest.registered_nodes[nodename]
	if def and def._transport_rel then
		player:add_pos(def._transport_rel)
	end
end
minetest.register_globalstep(function()
	for _, player in pairs(minetest.get_connected_players()) do
		checkplayer(player)
	end
end)
```

This creates "transporter" nodes that instantly zap me 16 nodes north or south.   I set pads made out of the stuff into the ground in my test world and walked back and forth between them and the effect worked fine, even with both server lag and artificial network lag.

I simulated up to ~1000ms of network lag using [my UDP delay proxy](https://gist.github.com/Warr1024/a93e5ade411d674ed0a51a73f7f5eaed)

File: src/network/networkprotocol.h
Comment (grorp) [inline src/network/networkprotocol.h:876]: Calling the new packet `TOCLIENT_PLAYER_POS` seems misleading because normal player position updates are not done via the new packet, but via `TOCLIENT_MOVE_PLAYER`.

Perhaps the new packet could be called `TOCLIENT_MOVE_PLAYER_REL` (relative)?

**Review by grorp — CHANGES_REQUESTED** [review]
This PR doesn't include backwards compatibility, i.e. old clients aren't affected by `player:add_pos` at all. Two approaches for backwards compatibility were suggested in #14023, please choose one and implement it.

File: src/network/networkprotocol.h
Comment (sfence) [reply src/network/networkprotocol.h]: Renamed.

**Comment by sfence** [comment]
@Warr1024 Thanks for the feedback.

@grorp All the mentioned problems should be fixed now.

File: games/devtest/mods/unittests/player.lua
Comment (grorp) [inline games/devtest/mods/unittests/player.lua:75]:
```suggestion
local function run_player_add_pos_tests(player)
```

to match the naming scheme of other functions in this file

File: games/devtest/mods/unittests/player.lua
Comment (grorp) [inline games/devtest/mods/unittests/player.lua:82]: Any reason for these specific values (`995 < newdist < 1001`)? You could use something like `assert(math.abs(vector.distance(pos, newpos) - 1000) < 1)` instead.

File: games/devtest/mods/unittests/player.lua
Comment (grorp) [inline games/devtest/mods/unittests/player.lua:85]:
```suggestion
unittests.register("test_player_add_pos", run_player_add_pos_tests, {player=true})
```

wrong copy-paste

**Review by grorp — CHANGES_REQUESTED** [review]
Thank you for the changes. Backwards compatibility works now.

In addition to my comments: Semicolons are unnecessary in Lua code, and please take a look at https://dev.minetest.net/Lua_code_style_guidelines.

File: games/devtest/mods/unittests/player.lua
Comment (sfence) [reply games/devtest/mods/unittests/player.lua]: Changed.

File: games/devtest/mods/unittests/player.lua
Comment (sfence) [reply games/devtest/mods/unittests/player.lua]: Fixed.

File: games/devtest/mods/unittests/player.lua
Comment (sfence) [reply games/devtest/mods/unittests/player.lua]: Fixed.

**Comment by sfence** [comment]
@grorp Should be fixed now.

I know about semicolons in Lua, but I am used to ending lines by it so much, that my hands sometimes add them automatically, without informing my brain about it :D.

So good to remind me. :)

File: games/devtest/mods/unittests/player.lua
Comment (grorp) [inline games/devtest/mods/unittests/player.lua:77]:
```suggestion
	player:add_pos(vector.new(0, 1000, 0))
```

File: games/devtest/mods/unittests/player.lua
Comment (grorp) [inline games/devtest/mods/unittests/player.lua:79]:
```suggestion
	player:add_pos(vector.new(0, -1000, 0))
```

File: games/devtest/mods/unittests/player.lua
Comment (grorp) [inline games/devtest/mods/unittests/player.lua:82]:
```suggestion
	assert(math.abs(newdist - 1000) <= 1)
```

File: games/devtest/mods/unittests/player.lua
Comment (grorp) [inline games/devtest/mods/unittests/player.lua:83]:
```suggestion
	assert(vector.distance(pos, backpos) <= 1)
```

**Review by grorp — CHANGES_REQUESTED** [review]
Alright, since my hint at Minetest's Lua style guide didn't work, I'll give concrete code style suggestions instead :) (according to https://dev.minetest.net/Lua_code_style_guidelines#Spaces)

Btw, I don't care about the style guide violations in `PlayerSAO::addPos` because the affected code is copied from the same file.

**Comment by sfence** [comment]
@grorp Some formatting fixes and PlayerSAO:addPos update to use `protocol_version` field in RemotePlayer class which I missed before because of looking for the field `proto_version`.

File: src/network/clientopcodes.cpp
Comment (sfan5) [inline src/network/clientopcodes.cpp:127]: suggestion: use one of the empty commands before this

File: src/network/clientpackethandler.cpp
Comment (sfan5) [inline src/network/clientpackethandler.cpp:645]:
```suggestion
	assert(player);
```

File: src/server/player_sao.cpp
Comment (sfan5) [inline src/server/player_sao.cpp]: surely this should be above the other check

File: src/network/clientopcodes.cpp
Comment (sfence) [reply src/network/clientopcodes.cpp:127]: Do you mean to change the packet type number from 0x64 to 0x65 and add info/comment that 0x64 is unused in `network protocol.h`?

File: src/network/clientpackethandler.cpp
Comment (sfence) [reply src/network/clientpackethandler.cpp]: fixed

File: src/server/player_sao.cpp
Comment (sfence) [reply src/server/player_sao.cpp]: fixed

Comment (grorp) [reply src/server/player_sao.cpp]: Since `setPos` also checks `isAttached`, the order doesn't actually matter.

**Review by grorp — APPROVED** [review]
LGTM if sfan5's comments are resolved

File: src/network/clientopcodes.cpp
Comment (sfan5) [reply src/network/clientopcodes.cpp:127]: no, I mean using for example `0x5d` since that's free

Comment (sfence) [reply src/network/clientopcodes.cpp]: Fixed.

**Comment by lhofhansl** [comment]
Code looks good. Have not tried it, though.

**Comment by lhofhansl** [comment]
This can be merged, right?

**Comment by sfence** [comment]
> This can be merged, right?

This PR is waiting for sfan5 approval, I think.

**Comment by sfan5** [comment]
You don't need to wait for my approval specifically, just check that my comments are reasonably addressed.

**Review by sfan5 — APPROVED** [review]
lgtm

**Pull request merged**
